src/data_types/strs.rs

```diff
@@ -151,4 +153,56 @@ impl CStr16 {
     pub fn to_u16_slice_with_nul(&self) -> &[u16] {
         unsafe { &*(&self.0 as *const [Char16] as *const [u16]) }
     }
+
+    /// Returns an iterator over this C string
+    pub fn iter<'a>(&'a self) -> CStr16Iter<'a> {
+        CStr16Iter {
+            inner: self,
+            pos: 0,
+        }
+    }
+}
+
+/// An iterator over `CStr16`.
+#[derive(Debug)]
+pub struct CStr16Iter<'a> {
+    inner: &'a CStr16,
+    pos: usize,
+}
+
+impl<'a> Iterator for CStr16Iter<'a> {
+    type Item = &'a Char16;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.pos >= self.inner.0.len() {
```

Comment:
...except if your implementation has an off-by-one, which I suspect it has. The last printable character of a CStr16 is the one at index len-2, but your loop will go up to the trailing NUL at index len-1.

Reply:
In this case, what I'm thinking about is that an `Iterator` should not interpret its inner data too much and it may be sad for someone trying to manually iterate over a `CStr16` and do some low-level operations but the iterator just simply skips the trailing NUL, so I pushed this check into the actual formatting process. Maybe this is not so good, however, because as far as I know `std::string` in C++11 ends with `\0` but its iterator does not return `\0`. I can't tell which one is better.

Reply:
I would personally propose not to expose the trailing NUL, for three reasons:

- It is redundant because the Iterator API already has None as a terminator.
- As you observed in this PR, emitting it means that many clients will need to account for its presence and handle it specially, because it is not like the other Char16s in the stream (owing to being in-band metadata rather than actual data).
- From the reverse perspective, I cannot think of any situation beyond memcpy where exposing it would help, because as soon as you need to combine multiple CStr16s, you need to have some sort of logic for dropping every NUL but the last one anyhow. And memcpy itself can be handled via lower-level access like the one provided by the to_u16_slice_* methods.

Reply:
I pushed a new commit and now `CStr16` skips the trailing NUL.